// @salesforce/core/dist/lib/project.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
const configFile_1 = require("./config/configFile");
const sfdxConfigAggregator_1 = require("./config/sfdxConfigAggregator");
const sfdxError_1 = require("./sfdxError");
const internal_1 = require("./util/internal");
const json_1 = require("./util/json");
/**
 * The sfdx-project.json config object. This file determines if a folder is a valid sfdx project.
 *
 * *Note:* Any non-standard (not owned by Salesforce) properties stored in sfdx-project.json should
 * be in a top level property that represents your project or plugin.
 *
 * @extends ConfigFile
 *
 * @example
 * const project = await SfdxProjectJson.retrieve<SfdxProjectJson>();
 * const myPluginProperties = project.get('myplugin') || {};
 * myPluginProperties.myprop = 'someValue';
 * project.set('myplugin', myPluginProperties);
 * await project.write();
 *
 * @see https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_ws_create_new.htm
 */
class SfdxProjectJson extends configFile_1.ConfigFile {
    static getFileName() {
        return internal_1.SFDX_PROJECT_JSON;
    }
    static getDefaultOptions(isGlobal = false) {
        const options = super.getDefaultOptions(isGlobal);
        options.isState = false;
        return options;
    }
    async read() {
        const contents = await super.read();
        // Verify that the configObject does not have upper case keys; throw if it does.  Must be heads down camel case.
        const upperCaseKey = json_1.findUpperCaseKeys(this.toObject());
        if (upperCaseKey) {
            throw sfdxError_1.SfdxError.create('@salesforce/core', 'core', 'InvalidJsonCasing', [upperCaseKey, this.getPath()]);
        }
        return contents;
    }
}
exports.SfdxProjectJson = SfdxProjectJson;
/**
 * Represents an SFDX project directory. This directory contains a {@link SfdxProjectJson} config file as well as
 * a hidden .sfdx folder that contains all the other local project config files.
 *
 * @example
 * const project = await Project.resolve();
 * const projectJson = await project.resolveProjectConfig();
 * console.log(projectJson.sfdxLoginUrl);
 */
class Project {
    /**
     * Do not directly construct instances of this class -- use {@link Project.resolve} instead.
     *
     * @private
     * @constructor
     */
    constructor(path) {
        this.path = path;
    }
    /**
     * Get a Project from a given path or from the working directory.
     * @param {string} path The path of the project.
     * @throws InvalidProjectWorkspace If the current folder is not located in a workspace.
     * @returns {Promise<Project>} The resolved project.
     */
    static async resolve(path) {
        return new Project(await this.resolveProjectPath(path));
    }
    /**
     * Performs an upward directory search for an sfdx project file.
     *
     * @param {string} [dir=process.cwd()] The directory path to start traversing from.
     * @returns {Promise<string>} The absolute path to the project.
     * @throws {SfdxError} **`{name: 'InvalidProjectWorkspace'}`** If the current folder is not located in a workspace.
     * @see fs.traverseForFile
     * @see {@link https://nodejs.org/api/process.html#process_process_cwd|process.cwd()}
     */
    static async resolveProjectPath(dir) {
        return internal_1.resolveProjectPath(dir);
    }
    /**
     * Returns the project path.
     * @returns {string}
     */
    getPath() {
        return this.path;
    }
    /**
     * Get the sfdx-project.json config. The global sfdx-project.json is used for user defaults
     * that are not checked in to the project specific file.
     *
     * *Note:* When reading values from {@link SfdxProjectJson}, it is recommended to use
     * {@link Project.resolveProjectConfig} instead.
     *
     * @param {boolean} isGlobal True to get the global project file, otherwise the local project config.
     */
    async retrieveSfdxProjectJson(isGlobal = false) {
        const prop = `sfdxProjectJson${isGlobal ? 'Global' : ''}`;
        const options = SfdxProjectJson.getDefaultOptions(isGlobal);
        if (!isGlobal) {
            options.rootFolder = this.getPath();
        }
        if (!this[prop]) {
            this[prop] = await SfdxProjectJson.retrieve(options);
        }
        return this[prop];
    }
    /**
     * The project config is resolved from local and global {@link SfdxProjectJson},
     * {@link SfdxConfigAggregator}, and a set of defaults. It is recommended to use
     * this when reading values from SfdxProjectJson.
     * @returns {object} A resolved config object that contains a bunch of different
     * properties, including some 3rd party custom properties.
     */
    async resolveProjectConfig() {
        if (!this.projectConfig) {
            // Get sfdx-project.json from the ~/.sfdx directory to provide defaults
            const global = await this.retrieveSfdxProjectJson(true);
            const local = await this.retrieveSfdxProjectJson();
            await global.read();
            await local.read();
            const defaults = {
                sfdcLoginUrl: 'https://login.salesforce.com'
            };
            this.projectConfig = _.defaults(local.toObject(), global.toObject(), defaults);
            // Add fields in sfdx-config.json
            _.assign(this.projectConfig, (await sfdxConfigAggregator_1.SfdxConfigAggregator.create()).getConfig());
            // LEGACY - Allow override of sfdcLoginUrl via env var FORCE_SFDC_LOGIN_URL
            if (process.env.FORCE_SFDC_LOGIN_URL) {
                this.projectConfig.sfdcLoginUrl = process.env.FORCE_SFDC_LOGIN_URL;
            }
        }
        return this.projectConfig;
    }
}
exports.Project = Project;
